src/snowflake/snowpark/_internal/debug_utils.py

```diff
@@ -249,21 +277,273 @@ def get_python_source_from_sql_error(top_plan: "SnowflakePlan", error_msg: str)
+                Used for proper tree connector formatting. Defaults to True.
+            file (Optional[file], optional): File handle to write output to. If None,
+                output is written to stdout. Defaults to None.
+
+        Returns:
+            None: This function prints output and doesn't return a value.
+
+        """
+        node = nodes[node_id]
+
+        connector = "└── " if is_last else "├── "
+
+        node_info = (
+            f"[{node['id']}] {node['type']} "
+            f"(In: {node['input_rows']:,}, Out: {node['output_rows']:,}, "
+            f"Mult: {node['row_multiple']:.2f}, Time: {node['exec_time']:.2f}%)"
+        )
+
+        self._write_output(f"{prefix}{connector}{node_info}", file)
+
+        extension = "    " if is_last else "│   "
+        new_prefix = prefix + extension
+
+        child_list = children.get(node_id, [])
+        for i, child_id in enumerate(child_list):
+            is_last_child = i == len(child_list) - 1
+            self.print_operator_tree(
+                nodes, children, child_id, new_prefix, is_last_child, file=file
+            )
+
+    def profile_query(
+        self,
+        query_id: str,
+        output_file: Optional[str] = None,
+        append_mode: bool = False,
+    ) -> None:
+        """
+        Profile a query and save the results to a file.
+
+        Args:
+            query_id: The query ID to profile
+            output_file: The file to save the results to. If not provided, the results will be printed to the console.
+
+        Returns:
+            None - output either to the console or to the file specified by output_file
+        """
+
+        stats_query = f"""
+            SELECT
+                operator_id,
+                operator_type,
+                operator_attributes,
+                operator_statistics:input_rows::number as input_rows,
+                operator_statistics:output_rows::number as output_rows,
+                CASE
+                    WHEN operator_statistics:input_rows::number > 0
+                    THEN operator_statistics:output_rows::number / operator_statistics:input_rows::number
+                    ELSE NULL
+                END as row_multiple,
+                execution_time_breakdown:overall_percentage::number as overall_percentage
+            FROM TABLE(get_query_operator_stats('{query_id}'))
+            ORDER BY step_id, operator_id
+            """
+        stats_connection = self.session._conn._conn.cursor()
+        stats_connection.execute(stats_query)
+        raw_results = stats_connection.fetchall()
+
+        column_names = [desc[0] for desc in stats_connection.description]
+        stats_result = [dict(zip(column_names, row)) for row in raw_results]
+
+        nodes, children, root_nodes = self.build_operator_tree(stats_result)
+        file_handle = None
+        if output_file:
+            mode = "a" if append_mode else "w"
+            file_handle = open(output_file, mode, encoding="utf-8")
```

Comment:
one helpful thing since you have a class now is that you can set `self.file_handle = open(...)` and have `self._write_output()` use it without explicitly passing file_handle each time.